### PR TITLE
fix alibi inference shapes for cached layer_past

### PR DESCRIPTION
**Pull request by sweinbach**
Restart the now reverted previous fix.

# History:
- Old PR was merged, after which some (small?) differences in model output became apparent in discussion with @sdtblck 
- Old PR was reverted
- This PR is opened to discuss the issue

# Tests and validations so far:

Inference was tested on a trained neox checkpoint (model size ~4B, 60k steps trained). Random sampling is deactivated (no top_k, top_p, temperature). 

1. **test NOT using recompute**, i.e. with cache values used in the text generation (interactive)
```
python deepy.py generate.py -d configs ... text_generation
``` 
(copy pasted from terminal)
Context prompt >>> Once upon a time
Generated Text: , the only way to get a job at any of the many colleges and universities in the United States of America was to have a high school diploma

=> the model is working
=> generated text seems far from random (though not a fairytale :-) )

2. **test using recompute**, i.e. with NO cache values used in the text generation (interactive) 
```
python deepy.py generate.py -d configs ... text_generation
``` 
(copy pasted from terminal)
Context prompt >>> Once upon a time
Generated Text: , the only way to get a job at any of the many colleges and universities in the United States of America was to have a high school diploma

=> The generated text is exactly the same as above

# Tests and validations that resulted in discussions.

The following line has been added to text_generation_utils.py to print logits.
image

3. **test NOT using recompute, logit output**

(copy pasted from terminal)
Context prompt >>> Once upon a time
generated_tokens tensor(15, device='cuda:0') generated_token_logits tensor(42.1250, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(301, device='cuda:0') generated_token_logits tensor(37.4375, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(1711, device='cuda:0') generated_token_logits tensor(33.6250, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(2333, device='cuda:0') generated_token_logits tensor(36.8750, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(327, device='cuda:0') generated_token_logits tensor(42.1875, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(1635, device='cuda:0') generated_token_logits tensor(40.0312, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(247, device='cuda:0') generated_token_logits tensor(39.0938, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(6834, device='cuda:0') generated_token_logits tensor(36.5000, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(554, device='cuda:0') generated_token_logits tensor(40.6250, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(1287, device='cuda:0') generated_token_logits tensor(34.9688, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(338, device='cuda:0') generated_token_logits tensor(36.5938, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(301, device='cuda:0') generated_token_logits tensor(39.9688, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(2329, device='cuda:0') generated_token_logits tensor(34.2500, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(45857, device='cuda:0') generated_token_logits tensor(33.4062, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(348, device='cuda:0') generated_token_logits tensor(40.5312, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(40126, device='cuda:0') generated_token_logits tensor(38.3125, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(280, device='cuda:0') generated_token_logits tensor(41.7188, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(301, device='cuda:0') generated_token_logits tensor(41.1875, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(5510, device='cuda:0') generated_token_logits tensor(36.5000, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(6405, device='cuda:0') generated_token_logits tensor(41.2812, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(338, device='cuda:0') generated_token_logits tensor(44.4062, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(7384, device='cuda:0') generated_token_logits tensor(43.4688, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(672, device='cuda:0') generated_token_logits tensor(45.3750, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(327, device='cuda:0') generated_token_logits tensor(46., device='cuda:0', dtype=torch.float16)
generated_tokens tensor(779, device='cuda:0') generated_token_logits tensor(39.9062, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(247, device='cuda:0') generated_token_logits tensor(42.6875, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(2556, device='cuda:0') generated_token_logits tensor(36.8438, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(3991, device='cuda:0') generated_token_logits tensor(42.8750, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(30480, device='cuda:0') generated_token_logits tensor(40.0625, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(17, device='cuda:0') generated_token_logits tensor(42.8125, device='cuda:0', dtype=torch.float16)
Generated Text: , the only way to get a job at any of the many colleges and universities in the United States of America was to have a high school diploma

=> the generated text is the same as above

4. **test using recompute, logit output**
(copy pasted from terminal)
Context prompt >>> Once upon a time
generated_tokens tensor(15, device='cuda:0') generated_token_logits tensor(42.1562, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(301, device='cuda:0') generated_token_logits tensor(37.4375, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(1711, device='cuda:0') generated_token_logits tensor(33.6250, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(2333, device='cuda:0') generated_token_logits tensor(36.9062, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(327, device='cuda:0') generated_token_logits tensor(42.1875, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(1635, device='cuda:0') generated_token_logits tensor(40.0625, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(247, device='cuda:0') generated_token_logits tensor(39.0938, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(6834, device='cuda:0') generated_token_logits tensor(36.5312, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(554, device='cuda:0') generated_token_logits tensor(40.6562, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(1287, device='cuda:0') generated_token_logits tensor(35.0312, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(338, device='cuda:0') generated_token_logits tensor(36.5625, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(301, device='cuda:0') generated_token_logits tensor(40.1250, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(2329, device='cuda:0') generated_token_logits tensor(34.2500, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(45857, device='cuda:0') generated_token_logits tensor(33.3125, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(348, device='cuda:0') generated_token_logits tensor(40.6562, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(40126, device='cuda:0') generated_token_logits tensor(38.2812, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(280, device='cuda:0') generated_token_logits tensor(41.6875, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(301, device='cuda:0') generated_token_logits tensor(41.1250, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(5510, device='cuda:0') generated_token_logits tensor(36.4688, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(6405, device='cuda:0') generated_token_logits tensor(41.3750, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(338, device='cuda:0') generated_token_logits tensor(44.3438, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(7384, device='cuda:0') generated_token_logits tensor(43.2500, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(672, device='cuda:0') generated_token_logits tensor(45.2188, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(327, device='cuda:0') generated_token_logits tensor(46.0625, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(779, device='cuda:0') generated_token_logits tensor(39.8750, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(247, device='cuda:0') generated_token_logits tensor(42.6562, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(2556, device='cuda:0') generated_token_logits tensor(36.8438, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(3991, device='cuda:0') generated_token_logits tensor(42.9375, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(30480, device='cuda:0') generated_token_logits tensor(39.9688, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(17, device='cuda:0') generated_token_logits tensor(42.9375, device='cuda:0', dtype=torch.float16)
Generated Text: , the only way to get a job at any of the many colleges and universities in the United States of America was to have a high school diploma

=> Generated text still the same
=> **Logits are different. This is the question at hand**

# Questions

- Why are logits different?
- Does the difference only occur for alibi? Is this a general issue if an issue at all?

### CODE REVIEW

**Comment by sweinbach** [comment]
I tested the generation using another model trained on rotary.

Context prompt >>> Once upon a time
generated_tokens tensor(15, device='cuda:0') generated_token_logits tensor(14.3281, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(246, device='cuda:0') generated_token_logits tensor(11.0156, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(577, device='cuda:0') generated_token_logits tensor(11.1484, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(531, device='cuda:0') generated_token_logits tensor(11.5938, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(9247, device='cuda:0') generated_token_logits tensor(11.5234, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(301, device='cuda:0') generated_token_logits tensor(14.6172, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(246, device='cuda:0') generated_token_logits tensor(11.8594, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(6665, device='cuda:0') generated_token_logits tensor(12.3047, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(15, device='cuda:0') generated_token_logits tensor(16.2969, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(308, device='cuda:0') generated_token_logits tensor(14.6250, device='cuda:0', dtype=torch.float16)
Generated Text: , a man was born to a woman,

<press enter to continue>

Context prompt >>> Once upon a time
generated_tokens tensor(15, device='cuda:0') generated_token_logits tensor(14.3359, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(246, device='cuda:0') generated_token_logits tensor(11.0234, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(577, device='cuda:0') generated_token_logits tensor(11.1484, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(531, device='cuda:0') generated_token_logits tensor(11.5938, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(9247, device='cuda:0') generated_token_logits tensor(11.5234, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(301, device='cuda:0') generated_token_logits tensor(14.6250, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(246, device='cuda:0') generated_token_logits tensor(11.8672, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(6665, device='cuda:0') generated_token_logits tensor(12.3047, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(15, device='cuda:0') generated_token_logits tensor(16.2969, device='cuda:0', dtype=torch.float16)
generated_tokens tensor(308, device='cuda:0') generated_token_logits tensor(14.6406, device='cuda:0', dtype=torch.float16)
Generated Text: , a man was born to a woman,

<press enter to continue>

**Comment by sweinbach** [comment]
As the problem is also occoring for rotary and not only alibi we could merge this one and open a dedicated bug? What do you think?

**Comment by sweinbach** [comment]
Trying to pin down the problem.

- I am able to reproduce on another system -> it's not the gpu
- Values are already different for mixed_x_layer (here: https://github.com/EleutherAI/gpt-neox/blob/19b16838b8275bac12b95ae2b84f1087041c6282/megatron/model/transformer.py#L434). I am comparing a baseline with recompute (two tokens) to the forward of a single token in question with a recompute inference (only second token with first token being in layer_past) using `delta = (baseline_recompute["mixed_x_layer"][1] - no_recompute["mixed_x_layer"][0]).abs().sum()`. This delta is positive. Note that only the second token in the baseline is compared.


Any idea why the output of this module changes? It should be independent of the input token count, right?

**Comment by sweinbach** [comment]
My conclusion so far:
- Torch does not guarantee the exact same output. There are deviations also dependent on input shape ([2, 1, 768] vs [1, 1, 768]). 
- Even ColumnParallelLinearChanged above which just applies F.linear runs into the same problems.
- This is probably within expected error boundaries

Would you agree? Any other opinion? It feels kind of strange.

**Comment by EricHallahan** [comment]
Does anything change if you [tell PyTorch to avoid deterministic algorithms](https://pytorch.org/docs/1.10.0/notes/randomness.html#avoiding-nondeterministic-algorithms)?

**Comment by sweinbach** [comment]
I used `torch.set_deterministic(True)``due to an older installed version of torch.

First failure for later reference:
RuntimeError: Deterministic behavior was enabled with either `torch.set_deterministic(True)` or `at::Context::setDeterministic(true)`, but this operation is not deterministic because it uses CuBLAS and you have CUDA >= 10.2. To enable deterministic behavior in this case, you must set an environment variable before running your PyTorch application: CUBLAS_WORKSPACE_CONFIG=:4096:8 or CUBLAS_WORKSPACE_CONFIG=:16:8. For more information, go to https://docs.nvidia.com/cuda/cublas/index.html#cublasApi_reproducibility

Then I ran with `CUBLAS_WORKSPACE_CONFIG=:16:8 python test_cp_linear.py`


Output 
```
[W Context.cpp:70] Warning: torch.set_deterministic is in beta, and its design and  functionality may change in the future. (function operator())
NeoXArgs.from_ymls() ['configs/small.yml', 'configs/local_setup.yml']
> initializing torch distributed ...
[2021-11-03 17:41:34,684] [INFO] [distributed.py:37:init_distributed] Not using the DeepSpeed or torch.distributed launchers, attempting to detect MPI environment...
[2021-11-03 17:41:34,709] [INFO] [distributed.py:89:mpi_discovery] Discovered MPI settings of world_rank=0, local_rank=0, world_size=1, master_addr=192.168.0.26, master_port=6000
[2021-11-03 17:41:34,709] [INFO] [distributed.py:47:init_distributed] Initializing torch distributed with backend: nccl
> initializing model parallel with size 1
MPU DP: [0]
MPU PP: [0]
MPU MP: [0]
> setting random seeds to 1234 ...
[2021-11-03 17:41:34,712] [INFO] [checkpointing.py:231:model_parallel_cuda_manual_seed] > initializing model parallel cuda seeds on global rank 0, model parallel rank 0, and data parallel rank 0 with model parallel seed: 3952 and data parallel seed: 1234
delta tensor(0.0006, device='cuda:0', dtype=torch.float16, grad_fn=<SumBackward0>)
delta2 tensor(0.0012, device='cuda:0', dtype=torch.float16, grad_fn=<SumBackward0>)
delta_changed tensor(0., device='cuda:0', dtype=torch.float16, grad_fn=<SumBackward0>)
```

**Comment by sweinbach** [comment]
Also removed the mp relevant code from the module in question. Here is a script independent of neox that reproduces the issue:

```
import torch
import torch.nn.functional as F

#torch.use_deterministic_algorithms(True)
torch.set_deterministic(True)

DTYPE = torch.half
DEVICE = "cuda:0"
HS = 256

torch.manual_seed(42)
torch.cuda.manual_seed(42)

class OwnLinear(torch.nn.Module):
    def __init__(self, input_size, output_size):
        super(OwnLinear, self).__init__()

        self.input_size = input_size
        self.output_size = output_size
     
        # weight
        self.weight = torch.nn.Parameter(torch.empty(
            self.output_size, self.input_size,
            device=DEVICE, dtype=DTYPE))
        torch.nn.init.normal_(self.weight, mean=0.0, std=0.02)

        # bias
        self.bias = torch.nn.Parameter(torch.empty(
            self.output_size,
            device=DEVICE,
            dtype=DTYPE))
       
        with torch.no_grad():
            self.bias.zero_()

    def forward(self, input_):
        return F.linear(input_, self.weight, self.bias)
       

#############################################
# Initialize

own_linear = OwnLinear(
            input_size=HS,
            output_size=3 * HS,
        ).to(DEVICE)

linear = torch.nn.Linear(HS, HS).to(DEVICE)
if DTYPE == torch.half:
    linear = linear.half()
#############################################
# create dummy input

baseline_input = torch.randn((2, 1, HS), dtype=DTYPE).to(DEVICE)
compare_input = baseline_input[1, :, :].view(1, 1, HS)

#############################################
# forward pass
own_baseline_out = own_linear(baseline_input)
own_compare_out = own_linear(compare_input)
delta = (own_baseline_out[1, :, :] - own_compare_out[0, :, :]).abs().sum()
print("own delta", delta, flush=True)

baseline_out = linear(baseline_input)
compare_out = linear(compare_input)
delta = (baseline_out[1, :, :] - compare_out[0, :, :]).abs().sum()
print("delta", delta, flush=True)
```
Called with

```
CUBLAS_WORKSPACE_CONFIG=:4096:8 python test_cp.py
```

Output
```
[W Context.cpp:70] Warning: torch.set_deterministic is in beta, and its design and  functionality may change in the future. (function operator())
own delta tensor(4.2915e-06, device='cuda:0', dtype=torch.float16,
       grad_fn=<SumBackward0>)
delta tensor(0.0002, device='cuda:0', dtype=torch.float16, grad_fn=<SumBackward0>)
```
- The difference between own linear and torch.nn.Linear now is the initialization. 
- The problem occurs both for both implementations
- I run a number of parameters. Not all result in a delta. It is most easy to get a delta for fp16
- Setting torch to deterministic does not make a difference
- There is a thread in the pytorch forum: https://discuss.pytorch.org/t/why-is-the-output-of-a-linear-layer-different-when-the-batch-size-is-1/93515/6

=>We are dealing with numeric issues
=> Let's merge again

**Comment by StellaAthena** [comment]
@sdtblck @EricHallahan What ever happened with this? Did nobody ever look at it?

**Comment by sweinbach** [comment]
@sdtblck @EricHallahan can any one of you review? friendly ping :-)

**Comment by sweinbach** [comment]
tldr / summary:

This is a straightforward change to respect  different shapes und cached inference with layer_past. The reason for not outputting the exact same scores is due to numerical issues. Backed by the discussion and the links above. 

I have run quite some inference with that code change.

**Review by StellaAthena — APPROVED** [review]
Looks good to me! Thanks for bearing with us while we got this approved.
